### PR TITLE
(#225) CMS Event tests

### LINKED ISSUE #225: CMS: Event tests
## As a cms user I want to be able to create Event content type to promote Event

> **ESTIMATE** TBD

### Acceptance Criteria

```gherkin
Feature: As a cms user I want to be able to create Event content type to promote Event


    Scenario: User is adding new event content type
        Given user is navigating to "/user/login"
        When user enters credentials
        And user clicks "Log in" button
        Then user is logged in and the user name "admin" is displayed in the toolbar
        And the tool bar appears at the top
        When user clicks on "Content" tab
        And user clicks on "Add content" action button
        And user clicks on "Event" content type
        Then page title is "Create Event"
        When user selects test site section
        And user fills out the following fields
            | fieldLabel               | value                                 | field_name                     |
            | Pretty URL               | test-event                            | field_pretty_url               |
            | Event Title              | Automated Test Event                  | title                          |
            | Browser Title            | Automated Test Event- Browser Title   | field_browser_title            |
            | Card Title               | Automated Test Event - Card Title     | field_card_title               |
            | Meta Description         | Automated Test Event Meta Description | field_page_description         |
            | Feature Card Description | Test Feature Card Description         | field_feature_card_description |
        And user types "e" in the "Event Series" autosuggest
        And user selects 1 event series from "Event Series" autosuggest
        And "Event Start Date & Time" date field is displaying todays date
        And "Event End Date & Time" date field is displaying todays date
        And user enters "101000P" in the "Event End Date & Time" time field
        And user fills out the following fields
            | fieldLabel         | value              | field_name       |
            | Venue              | NCI Shady Grove    | field_venue      |
            | Room               | 2E626              | field_room       |
            | City/State/Country | Rockville, MD, USA | field_city_state |
        And user fills out "Body" text area with "This is a description of event content type."
        And user selects 1 Lead Image from the list of images
        And user selects 2 Promo Image from the list of images
        And user remembers the source of selected lead image for further verification
        And user selects "Published" from "Save as" dropdown
        Then user saves the content page

    Scenario: Verify newly created content
        Given user is navigating to the front end site with path site section plus "test-event"
        Then page title is "Automated Test Event"
        And the Event Series title matches selected event series
        And the event date is displaying todays date
        And the event start time displays in the following format "([0-9]|[0-9][0-9]):[0-9][0-9] (AM|PM)"
        And the event end time displays "10:00 PM"
        And the "venue" location field displays "NCI Shady Grove"
        And the "room" location field displays "2E626"
        And the "city-state" location field displays "Rockville, MD, USA"
        And "Add to Outlook Calendar" is a link with href that matches "\/node\/\d+\/icalendar"
        And description reads "This is a description of event content type."
        And the lead image for event is matching earlier selected image
 And the page contains meta tags with the following names
            | name        | content                               |
            | description | Automated Test Event Meta Description |

    Scenario: Verify the "all day event" checkbox, edit and republish
        Given user is navigating to "/user/login"
        When user enters credentials
        And user clicks "Log in" button
        Then user is logged in and the user name "admin" is displayed in the toolbar
        And the tool bar appears at the top
        When user clicks on "Content" tab
        And user clicks on title with url "test-event" from the list of content
        And user clicks on the tool bar status green button "Published"
        And user clicks "View in edit form" button from other actions
        And user checks "All Day Event?" checkbox
        When user saves the content page
        And user clicks on the tool bar status green button "Editing"
        And user selects "Quick Publish" from workflow actions

    Scenario: Verify the "all day event" checkbox hides time on front end
        Given user is navigating to the front end site with path site section plus "test-event"
        Then page title is "Automated Test Event"
        And the event date displays today date
        And event date does not display time


    Scenario: Edit and republish event content type
        Given user is navigating to "/user/login"
        When user enters credentials
        And user clicks "Log in" button
        Then user is logged in and the user name "admin" is displayed in the toolbar
        And the tool bar appears at the top
        When user clicks on "Content" tab
        And user clicks on title with url "test-event" from the list of content
        And user clicks on the tool bar status green button "Published"
        And user clicks "View in edit form" button from other actions
        And user clears out "Pretty URL" field
        And user clears out "Event Title" field
        And user clears out "Card Title" field
        And user clears out "Meta Description" field
  And user clears out "Venue" field
        And user clears out "City/State/Country" field
        And user clears out "Room" field
        And user fills out the following fields
            | fieldLabel         | value                                        | field_name             |
            | Pretty URL         | test-event-edited                            | field_pretty_url       |
            | Event Title        | Automated Test Event Edited                  | title                  |
            | Card Title         | Automated Test Event - Card Title Edited     | field_card_title       |
            | Meta Description   | Automated Test Event Meta Description Edited | field_page_description |
            | Venue              | NCI Shady Grove Edited                       | field_venue            |
            | Room               | 2E626 Edited                                 | field_room             |
            | City/State/Country | Arlington, VA, USA                           | field_city_state       |
        And user clears out "Event Series" field
        And user types "e" in the "Event Series" autosuggest
        And user selects 2 event series from "Event Series" autosuggest
       And user checks "All Day Event?" checkbox
        And user enters tomorrows date in "Event End Date & Time" date field
        And user removes the Lead Image
        And user selects 2 Lead Image from the list of images
        And user remembers the source of selected lead image for further verification
        When user saves the content page
        And user clicks on the tool bar status green button "Editing"
        And user selects "Quick Publish" from workflow actions

    Scenario: Verify edited content
        Given user is navigating to the front end site with path site section plus "test-event-edited"
        Then page title is "Automated Test Event Edited"
        And the Event Series title matches selected event series
        And the event date displays today as a start and tomorrow as the end of event
        And event date does not display time
        And the "venue" location field displays "NCI Shady Grove Edited"
        And the "room" location field displays "2E626 Edited"
        And the "city-state" location field displays "Arlington, VA, USA"
        And "Add to Outlook Calendar" is a link with href that matches "\/node\/\d+\/icalendar"
        And description reads "This is a description of event content type."
        And the lead image for event is matching earlier selected image
        And the page contains meta tags with the following names
            | name        | content                                      |
            | description | Automated Test Event Meta Description Edited |


    Scenario: Clean up
        Given user is navigating to "/user/login"
        When user enters credentials
        And user clicks "Log in" button
        Then user is logged in and the user name "admin" is displayed in the toolbar
        And the tool bar appears at the top
        When user clicks on "Content" tab
        And user selects a checkbox next to title with url "test-event-edited" from the list of content
        And user clicks on "Apply to selected items" content action button
        Then page title is "Are you sure you want to delete this content item?"
        When user clicks on "Delete" button
        Then the confirmation text "Deleted 1 content item." appears on a screen
        And the content item with url "test-event-edited" does not exist in the list of content
```

### Sub-Tasks
- [ ] Create Component
- [ ] Implement Unit Tests for Component

### Notes

*Some complementary notes if necessary:*
* > Here goes a quote from an email
* Here goes whatever useful information can exist…
* More notes, or technical nuances/direction

### PR DESCRIPTION
**Pull request by alinai**
Closes #225 
Added Feature File
Added some tests

### CODE REVIEW

**Comment by alinai** [comment]
Test run result:
image

**Comment by alinai** [comment]
Cypress Test run result on dev-ac:
image

**Comment by alinai** [comment]
@KateMashkinaNIH , please review. Thanks!

File: cypress/e2e/CMS/$Event/index.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+});
+
+And('{string} is a link with href that matches {string}', (linkText, linkHref) => {
+    const regex = new RegExp(linkHref);
+    cy.get('div#cgvBody p').find('a').should('have.text', linkText);
+    cy.get('div#cgvBody p').find('a').should('have.attr', 'href').and('match', regex);
+});
+
+And('description reads {string}', (desc) => {
+    cy.get('#cgvBody > :nth-child(6)').should('be.visible')
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$Event/index.js:123]: cy.get(``p:contains(${desc})``).should('be.visible')

File: cypress/e2e/CMS/$Event/index.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+And('event date does not display time', () => {
+    cy.get('p.event-date-time').should('not.include.text', '|');
+});
+
+And('user removes the Lead Image', () => {
+    cy.get('#edit-field-image-article-current-items-0-remove-button').click({ force: true })
+});
+
+And('the event date displays today as a start and tomorrow as the end of event', () => {
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$Event/index.js:148]: remove this step since we got rid of it in a feature file

**Review by KateMashkinaNIH — CHANGES_REQUESTED** [review]

File: cypress/e2e/CMS/$Event/index.js [hunk 3]
```diff
@@ -0,0 +1,177 @@
+/// <reference types="Cypress" />
+import { And } from 'cypress-cucumber-preprocessor/steps';
+
+
+And('user types {string} in the {string} autosuggest', (value, fieldLabel) => {
+    cy.get(`div[class*="field-event-series"] label:contains("${fieldLabel}")`).type(value);
+    cy.wait(200);
```
[the remark this answers is not in the document]
Comment (alinai) [reply cypress/e2e/CMS/$Event/index.js:7]: wait is removed

File: cypress/e2e/CMS/$Event/index.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+});
+
+And('user selects {int} event series from {string} autosuggest', (position, fieldLabel) => {
+    cy.get('#ui-id-1 li').eq(position - 1).click({ force: true });
```
[the remark this answers is not in the document]
Comment (alinai) [reply cypress/e2e/CMS/$Event/index.js:11]: done

File: cypress/e2e/CMS/$Event/index.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        imageSrc = $el[0].getAttribute('src')
+    });
+});
+
+And('the Event Series title matches selected event series', () => {
+    let textSrc;
+    let expectedText;
+    cy.get('#cgvBody h2').then($el => {
+        textSrc = $el[0].innerText;
+        expectedText = textSrc.trim();
```
[the remark this answers is not in the document]
Comment (alinai) [reply cypress/e2e/CMS/$Event/index.js:63]: done

File: cypress/e2e/CMS/$Event/index.js (hunk 1 above)
Comment (alinai) [reply cypress/e2e/CMS/$Event/index.js:123]: done

File: cypress/e2e/CMS/$Event/index.js (hunk 2 above)
Comment (alinai) [reply cypress/e2e/CMS/$Event/index.js:148]: removed

**Comment by alinai** [comment]
Cypress test run result on dev-ac:
image

**Comment by alinai** [comment]
npm test run result on dev-acsf:

image

**Comment by alinai** [comment]
@KateMashkinaNIH , please review again. Thanks!

**Comment by KateMashkinaNIH** [comment]
@alinai step And Event Start Date & Time date field is displaying todays date is failing due to the expected date difference - if the day and month has 1 digit, then it should have a 0 appended ( 2023-01-03 ) - currently it builds expected string as 2023-1-3
Screen Shot 2023-01-03 at 9 42 06 AM

**Comment by alinai** [comment]
Cypress test run on dev-ac:
image
npm test run on dev-acsf:
image

**Comment by alinai** [comment]
npm test run on dev-acsf:
image

**Review by KateMashkinaNIH — APPROVED** [review]

**Pull request merged**